completion_time_analysis.py

```diff
@@ -0,0 +1,237 @@
+        return x if x.tzinfo else x.replace(tzinfo=timezone.utc)
+    ts = pd.to_datetime(str(x), utc=True, errors="coerce")
+    if pd.isna(ts):
+        return None
+    return ts.to_pydatetime()
+
+def _created_at(issue: Issue) -> Optional[datetime]:
+    """Get created date from issue."""
+    return _to_utc_dt(issue.created_date) if issue.created_date else None
+
+def _updated_at(issue: Issue) -> Optional[datetime]:
+    """Get updated date from issue."""
+    return _to_utc_dt(issue.updated_date) if issue.updated_date else None
+
+def _closed_at_from_events(issue: Issue) -> Optional[datetime]:
+    """Find closed date from events."""
+    cands = []
+    for e in issue.events:
+        if e.event_type and e.event_type.lower() == "closed" and e.event_date:
+            ts = _to_utc_dt(e.event_date)
+            if ts:
+                cands.append(ts)
+    return max(cands) if cands else None
+
+def _closed_at(issue: Issue) -> Optional[datetime]:
+    """Get closed date from issue events, or fallback to updated_date if closed."""
+    ts = _closed_at_from_events(issue)
+    if ts:
+        return ts
+    # Fallback: if issue is closed, use updated_date
+    if issue.state == State.closed:
+        return _updated_at(issue)
+    return None
+
+def _labels(issue: Issue) -> List[str]:
+    """Get labels from issue."""
+    return issue.labels if issue.labels else ["unlabeled"]
+
+def _url(issue: Issue) -> Optional[str]:
+    """Get URL from issue."""
+    return issue.url if issue.url else f"https://github.com/python-poetry/poetry/issues/{issue.number}" if issue.number >= 0 else None
+
+# -------------------------- main analysis --------------------------------- #
+
+class CompletionAnalysis:
+    """
+    CLOSED issues only. One chart + concise summary.
+    """
+
+    def __init__(self):
+        self.issues: List[Issue] = DataLoader().get_issues()
+        self.user_filter: Optional[str] = config.get_parameter("user")
+        self.label_filter: Optional[str] = config.get_parameter("label")
+        self.since: Optional[str] = config.get_parameter("since")
+        self.filtered_issues = self._filter_issues()
+
+    def _filter_issues(self) -> List[Issue]:
+        items = self.issues
+        if self.user_filter:
+            items = [i for i in items if i.creator == self.user_filter]
+        if self.label_filter:
+            items = [i for i in items if self.label_filter in _labels(i)]
+        if self.since:
+            start = pd.to_datetime(self.since, utc=True, errors="coerce")
+            if not pd.isna(start):
+                items = [i for i in items if (c := _created_at(i)) and c >= start.to_pydatetime()]
+        return items
+
+    def _completion_days(self, issue: Issue) -> Optional[float]:
+        if issue.state != State.closed:
+            return None
+        c0 = _created_at(issue)
+        c1 = _closed_at(issue)
+        if not c0 or not c1:
+            return None
+        d = (c1 - c0).total_seconds() / 86400.0
```

Comment:
optional simplification
```suggestion
        d = (c1 - c0).days
```

Reply:
changes applied